nav2_costmap_2d/plugins/costmap_filters/keepout_filter.cpp

```diff
@@ -0,0 +1,186 @@
+}
+
+void KeepoutFilter::process(
+  nav2_costmap_2d::Costmap2D & master_grid,
+  int min_i, int min_j, int max_i, int max_j,
+  double /*robot_x*/, double /*robot_y*/, double /*robot_yaw*/)
+{
+  if (!map_filter_) {
+    curr_time_ = node_->now();
+    // Show warning message every 2 seconds to not litter an output
+    if (curr_time_ - prev_time_ >= rclcpp::Duration(std::chrono::milliseconds(2000))) {
+      RCLCPP_WARN(node_->get_logger(), "Map filter was not received");
+      prev_time_ = curr_time_;
+    }
+    return;
+  }
+
+  unsigned int min_i_u, min_j_u, max_i_u, max_j_u;  // Unsigned versions of window bounds
+  double wx, wy;  // world coordinates
+
+  // unsigned<-signed conversions.
+  // All map bounds are positive, so this will be safe.
+  min_i_u = (unsigned int)min_i;
+  min_j_u = (unsigned int)min_j;
+  max_i_u = (unsigned int)max_i;
+  max_j_u = (unsigned int)max_j;
+
+  // Optimization: iterate only by map-filter window bounds corresponding to
+  // (min_i, min_j)..(max_i, max_j) window.
+  // ToDo: after costmap rotation will be added, this should be re-worked.
+  int mf_tmp_x, mf_tmp_y;  // Temp signed map_filter_ coordinates using in calculations
+  unsigned int mf_min_x_u, mf_min_y_u, mf_max_x_u, mf_max_y_u;  // map_filter_ bounds
+
+  // Calculating map_filter_ bounds corresponding to (min_i, min_j) corner
+  mapToWorld(min_i_u, min_j_u, wx, wy);
+  map_filter_->worldToMapNoBounds(wx, wy, mf_tmp_x, mf_tmp_y);
+  mf_tmp_x = std::max(mf_tmp_x, 0);
+  mf_tmp_y = std::max(mf_tmp_y, 0);
+  mf_min_x_u = (unsigned int)mf_tmp_x;  // unsigned<-signed conversion
+  mf_min_y_u = (unsigned int)mf_tmp_y;  // unsigned<-signed conversion
+  // Checking that mf_min_x_u/mf_min_y_u are not out of map_filter_ bounds
+  if (mf_min_x_u > map_filter_->getSizeInCellsX() || mf_min_y_u > map_filter_->getSizeInCellsY()) {
+    return;
+  }
+
+  // Calculating map_filter_ bounds corresponding to (max_i, max_j) corner
+  mapToWorld(max_i_u, max_j_u, wx, wy);
+  map_filter_->worldToMapNoBounds(wx, wy, mf_tmp_x, mf_tmp_y);
+  // Checking that max_x/max_y are not out of map_filter_ bounds
+  if (mf_tmp_x < 0 || mf_tmp_y < 0) {
+    return;
+  }
+  mf_max_x_u = (unsigned int)mf_tmp_x;  // unsigned<-signed conversion
+  mf_max_y_u = (unsigned int)mf_tmp_y;  // unsigned<-signed conversion
+  mf_max_x_u = std::min(mf_max_x_u, map_filter_->getSizeInCellsX());
+  mf_max_y_u = std::min(mf_max_y_u, map_filter_->getSizeInCellsY());
+
+  unsigned int i, j;  // map_filter_ iterators
+  unsigned int mx, my;  // costmap_ coordinates
+  unsigned char data;  // map_filter_ element data
+
+  for (i = mf_min_x_u; i < mf_max_x_u; i++) {
+    for (j = mf_min_y_u; j < mf_max_y_u; j++) {
+      data = (*map_filter_)[map_filter_->getIndex(i, j)];
+      if (data != nav2_util::OCC_GRID_UNKNOWN) {
+        // Converting each point on map_filter to world coordinates
+        map_filter_->mapToWorld(i, j, wx, wy);
```

Comment:
This filter derives from the costmap filter layer which derives from a costmap layer which has a `costmap_` to store data. Why keep the `map_filter_` around when you have a perfectly good `costmap_` you're not making use of (but in resetting and matching size, you're allocating, so you're doubling the memory without using half of it)

Reply:
So, am I correctly understand, that you are suggesting to iterate over `costmap_` indexes inside the window, get world coordinates for them and then convert these coordinates back to `mask_`'s (formerly `map_filter_`) indexes?
If so, there only one routine should be copy-pasted from Costmap2D API: `worldToMap` which will be used for `mask_`.

Reply:
What I'm suggesting is when you get the map from the callback from the map server, loading that into `costmap_` rather than saving it to `map_filter_`. This is to be consistent with other plugins, but also because right now every time you do `resetMaps()` or `matchSize()` its meaningless because you don't use the `costmap_` that's being modified. Also you're allocating an entirely unused `costmap_` of size `size_x` by `size_y` (which for a filter of a full mask of the map is a complete memory duplicate but empty)

Reply:
All `updateWithMax()/updateWithOverride()/etc...` methods require that `costmap_` should have the same sizes, scale and origin as `master_grid`. This is not guaranteed by the `mask_`: map mask might have smaller or larger scale, origin and (in the future) - orientation. Also, I can not understand how can we store `mask_` into `costmap_` for rolling windows with dynamically changing origin of the `costmap_` and `master_grid`? So, I think, we can not directly store `mask_` into the `costmap_`.

Therefore, we need to have space conversions from `mask_` indexes -> to world coordinates -> to `costmap_` indexes, that are currently discussing. There is no mapToWorld/worldToMap conversion API for OccupancyGrid-s, so we may need to store an additional `mask_costmap_` with Costmap2D type, or using directly OccupancyGrid, but with API wrapper that I've made in `nav2_utils/occupancy_grid.hpp`. In my opinion, second option is better in terms of performance, but this is under discussion.

Reply:
tl;dr from slack message:
- You don't have to use the updateWithXYZ() helpers if you don't use `costmap_` then, you can just set that logic yourself and use `mask_`. Then you can stop allocating `costmap_` and leave it empty
 - I don't think this PR as-is deals with different sized masks from the map. If that's wrong, let me know and show a toy demo of it and we can move on on that point

Reply:
Though I suppose you could use the nav2_msgs/Costmap if you liked. We'd have to think a little if that made sense in the differences between them

Reply:
Okay, It looks like I've got your point.
The main item is that I will going to use - is that make `costmap_` with 0-size and do not use it at all.
When the `mask_` message arrived in the call-back I will allocate the same sized Costmap2D object, say `mask_costmap_`. And in the process there is possible to calculate `mask_costmap_` and `master_grid` interlacing square area coordinates. Using 2-nested cycle by `mask_costmap_` in interlaced area I can directly update `master_grid` without using a costmap_  at all.

> I don't think this PR as-is deals with different sized masks from the map. If that's wrong, let me know and show a toy demo of it and we can move on on that point. But I agree with your reasoning that having that be true would not allow us to use the costmap_ variable - see point 1.

I've checked mask maps with different from origins, not sized, but they are worked well. I do not think the size is a principal difference in current implementation: it should work as well. All linear transforms (except of rotation) are already supported for map masks and this algorithm. E.g. the first code revision, https://github.com/ros-planning/navigation2/pull/1882/commits/21944a9cbedddd50581f88d1966ba7196f5cae4f#diff-f67a16397bb5979b7cbf4729ea53a451R119-R137 did that: the algorithm is independent on `mask_` (formerly `map_filter_ `) size, scale or position. I will make an example later during the commit stage, as you've requested for rolling (used in local planner) and static (used in global planner) costmaps.